presto-orc/src/main/java/com/facebook/presto/orc/metadata/StripeMetaCache.java

```diff
@@ -0,0 +1,88 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.metadata;
+
+import com.facebook.airlift.log.Logger;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+
+import java.util.List;
+
+import static com.facebook.presto.orc.metadata.StripeMetaCacheMode.FOOTER;
+import static com.facebook.presto.orc.metadata.StripeMetaCacheMode.INDEX;
+import static com.facebook.presto.orc.metadata.StripeMetaCacheMode.INDEX_AND_FOOTER;
+import static com.facebook.presto.orc.metadata.StripeMetaCacheMode.NONE;
+import static java.util.Objects.requireNonNull;
+
+public class StripeMetaCache
+{
+    private static final Logger log = Logger.get(StripeMetaCache.class);
+    private final List<Integer> cacheOffsets;
+    private final StripeMetaCacheMode cacheMode;
+    private final Slice data;
+
+    public StripeMetaCache(List<Integer> offsets, StripeMetaCacheMode mode, Slice data)
+    {
+        this.cacheOffsets = ImmutableList.copyOf(requireNonNull(offsets, "offsets is null"));
+        this.data = requireNonNull(data, "data is null");
+        this.cacheMode = filterMode(requireNonNull(mode, "mode is null"));
+    }
+
+    // work only with supported modes
+    private StripeMetaCacheMode filterMode(StripeMetaCacheMode mode)
+    {
+        if (mode == INDEX || mode == FOOTER || mode == INDEX_AND_FOOTER) {
+            return mode;
+        }
+        return NONE;
+    }
+
+    public Slice getStripeFooterSlice(int stripeOrdinal, long sliceLength)
+    {
+        return getSlice(stripeOrdinal, FOOTER, sliceLength);
+    }
+
+    public Slice getStripeIndexSlice(int stripeOrdinal)
+    {
+        return getSlice(stripeOrdinal, INDEX, -1);
+    }
+
+    private Slice getSlice(int stripeOrdinal, StripeMetaCacheMode kind, long expectedSliceLength)
+    {
+        if (!cacheMode.has(kind)) {
+            return null;
+        }
+
+        int index = -1;
+        if (cacheMode == kind) {
+            index = stripeOrdinal;
+        }
+        else if (cacheMode == INDEX_AND_FOOTER) {
+            index = stripeOrdinal * 2 + (kind == INDEX ? 0 : 1);
+        }
+
+        // cache might contain data only for first N stripes
+        if (index >= 0 && index < cacheOffsets.size() - 1) {
+            int dataLength = cacheOffsets.get(index + 1) - cacheOffsets.get(index);
+            if (dataLength != expectedSliceLength && expectedSliceLength != -1) {
+                log.warn("Size mismatch between expected %s and cached data %s for stripe %s", expectedSliceLength, dataLength, stripeOrdinal);
+                return null;
+            }
+
+            return data.slice(cacheOffsets.get(index), dataLength);
+        }
+
+        return null;
```

Comment:
In presto code base, returning nulls are discouraged. It has to be optional<>

Reply:
Done